spock/src/main/java/org/jboss/weld/spock/impl/EagerExceptionRenderer.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2018, Red Hat, Inc., and individual contributors
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.weld.spock.impl;
+
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.spockframework.runtime.SpockAssertionError;
+import org.spockframework.runtime.extension.IGlobalExtension;
+import org.spockframework.runtime.model.FeatureInfo;
+import org.spockframework.runtime.model.SpecInfo;
+
+/**
+ * A global Spock extension that eagerly renders exceptions, otherwise {@code toString()} might be called
+ * on a CDI proxy when the container is shut down already and the call cannot be done.
+ *
+ * <p>If in a power assertion output you see something like
+ * {@code Something$Proxy$_$$_WeldClientProxy@4303056f (renderer threw IllegalStateException)},
+ * this extension might miss some places where to eagerly render the exception, and you should
+ * open an issue with a reproducer.
+ *
+ * @author Björn Kautler
+ */
+public class EagerExceptionRenderer implements IGlobalExtension {
```

Comment:
No, the assertion error is thrown in time.
Otherwise I could also not catch it there.
The problem is, that Spock remembers the values of assertion expressions and if the assertions failed and the error is printed, it prints power assertion output where the value of each part of the expression is printed.
For most expressions, this would then call `toString()`, but only at the time the exception is actually printed.
At **that** time the Weld container is long shut down usually and you would get `IllegalStateExceptions` when the Weld client proxy is asked for `toString()` and the container is already gone.
The exception caches the result of the rendering, so this extension triggers the rendering while the container is still alive so that the rendering works properly which is then cached for later by Spock.

For an example see the last section of the readme.